src/memcpy_vfp.c

```diff
@@ -0,0 +1,151 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "driver.h"
+
+#define BLOCK_SIZE  4096
+
+static char tmp_buf0[BLOCK_SIZE] __attribute__((aligned (128)));
+
+static inline void tegra_aligned_block_copy_vfp(void *dst, void *src, int size)
+{
+    asm volatile(
+        "   .fpu vfpv3-d16          \n\t"
+        "   .arch armv7a            \n\t"
+        "0:                         \n\t"
+        "   pld   [%1, #128]        \n\t"
+        "   pld   [%1, #160]        \n\t"
+        "   pld   [%1, #192]        \n\t"
+        "   pld   [%1, #224]        \n\t"
+        "   pld   [%1, #256]        \n\t"
+        "   vldm  %1!, {d0-d15}     \n\t"
+        "   subs  %2, %2, #128      \n\t"
+        "   vstm  %0!, {d0-d15}     \n\t"
+        "   bgt   0b                \n\t"
+        : "+r" (dst), "+r" (src), "+r" (size)
+        :
+        : "cc", "d8", "d9", "d10", "d11", "d12", "d13", "d14", "d15");
+}
+
+void tegra_copy_block_vfp(char *dst, char *src, int size,
+                          bool bounce, bool move)
+{
+    int i, dir, block_size = BLOCK_SIZE;
+    char *tmpbuf = tmp_buf0;
+    char *psrc = src;
+    char *pdst = dst;
+
+    assert(((uintptr_t)src & 127) == 0);
+    assert(((uintptr_t)dst & 127) == 0);
+    assert((size & 127) == 0);
+
+    do {
+        if (size < block_size || !bounce) {
+            block_size = size;
+            move = false;
+        }
+
+        dir = (pdst > psrc && move) ? -1 : 1;
+
+        if (dir < 0) {
+            psrc += size - block_size;
+            pdst += size - block_size;
+        }
+
+        for (i = 0; i < size / block_size; i++) {
+            if (!bounce)
+                tmpbuf = pdst;
+
+            tegra_aligned_block_copy_vfp(tmpbuf, psrc, block_size);
+
+            if (bounce)
+                tegra_aligned_block_copy_vfp(pdst, tmpbuf, block_size);
+
+            psrc += block_size * dir;
+            pdst += block_size * dir;
+        }
+
+        size -= block_size * i;
+        pdst = max(pdst, dst);
+        psrc = max(psrc, src);
+
+    } while (size);
+}
+
+void tegra_do_memcpy_vfp(void *dst, const void *src, int size, bool bounce)
+{
+    bool src_aligned = !((uintptr_t)src & 127);
+    bool dst_aligned = !((uintptr_t)dst & 127);
+    int left = size;
+    int block_size;
+    char *csrc = (char *)src;
+    char *cdst = dst;
+    char *pdst;
+
+    if (src_aligned && dst_aligned && left > 127) {
+        tegra_copy_block_vfp(cdst, csrc, size & ~127, bounce, false);
+        csrc += size & ~127;
+        cdst += size & ~127;
+        left = size & 127;
+    }
+
+    if (left > 127) {
+        uintptr_t addr, addr_aligned, offset;
+
+        /* having dst bouncing instead of src is always more efficient */
+        addr         = (uintptr_t)csrc;
+        addr_aligned = TEGRA_ALIGN(addr, 128);
+        offset       = addr_aligned - addr;
```

Comment:
These three lines are a roundabout way of writing "offset = 128 - (uintptr_t)csrc & 127;", I think...

Reply:
Thanks :+1: